npep/npep-126-egress-traffic-control.md

```diff
@@ -76,13 +76,249 @@ selected cluster workloads to k8s-apiservers for securing the server.
+}
+```
+
+Note that `AdminNetworkPolicyPeer` will changed to
+`AdminNetworkPolicyEgressPeer` and `AdminNetworkPolicyIngressPeer` since ingress and
+egress peers have started to diverge at this point and it is easy to
+maintain it with two sets of peer definitions.
+This ensures nodes can be referred to only as "egress peers".
+
+Example: Admin wants to deny egress traffic from tenants who don't have
+`restricted`, `confidential` or `internal` level security clearance
+to control-plane nodes at 443 and 6443 ports in the cluster
+
+```
+apiVersion: policy.networking.k8s.io/v1alpha1
+kind: AdminNetworkPolicy
+metadata:
+  name: node-as-egress-peers
+spec:
+  priority: 55
+  subject:
+    namespaces:
+      matchExpressions:
+        - {key: security, operator: notIn, values: [restricted, confidential, internal]}
+  egress:
+  - name: "deny-all-egress-to-kapi-server"
+    action: "Deny"
+    to:
+    - nodes:
+        matchLabels:
+          node-role.kubernetes.io/control-plane: true
+    ports:
+      - portNumber:
+          protocol: TCP
+          port: 443
+      - portNumber:
+          protocol: TCP
+          port: 6443
+```
+
+### Implementing egress traffic control towards external destinations
+
+This NPEP proposes to add a new type of `AdminNetworkPolicyEgressPeer` called `ExternalNetworks`
+to be able to explicitly select external destinations (based on the `externalNetworkSet`'s
+labels) in the cluster. This peer type will not be supported in `AdminNetworkPolicyIngressPeer`.
+
+```
+// AdminNetworkPolicyEgressPeer defines a peer to allow traffic to.
+// Exactly one of the selector pointers must be set for a given peer. If a
+// consumer observes none of its fields are set, they must assume an unknown
+// option has been specified and fail closed.
+// +kubebuilder:validation:MaxProperties=1
+// +kubebuilder:validation:MinProperties=1
+type AdminNetworkPolicyEgressPeer struct {
+    <snipped>
+	// ExternalNetworks defines a way to select ExternalNetworkSets
+	// that consist of network CIDRs that live outside the cluster as a peer.
+	// This field follows standard label selector semantics; if present
+	// but empty, it selects all ExternalNetworkSets defined in the cluster.
+	//
+	// Support: Core
+	//
+	// +optional
+	ExternalNetworks *metav1.LabelSelector `json:"externalNetworks,omitempty"`
+}
+```
+
+An `externalNetworkSet` is a new object used to define a set of networks outside
+the cluster. This is defined as a new object so that its easy for users to group
+their external entities and then refer to that group from the different egress rules.
+This ensures consistency without the users having to manually change the ANP/BANP
+to accommodate the changing needs of the network cidrs. They can directly edit the
+new CRD which guarantees all rules effecting that peer will take effect automatically.
+In the future when we define `DeveloperNetworkPolicies`, this will come even more handy
+for uniformity specially across `Pass` actions.
```

Comment:
```suggestion
An `ExternalNetworkSet` is a new object used to define a set of networks outside
the cluster. This is defined as a new object so that it's easy for users to group
their external entities and then refer to that group from the different egress rules.
This ensures consistency without the users having to manually change the ANP/BANP
to accommodate the changing needs of the network CIDRs. They can directly edit the
new object which guarantees all rules affecting that peer will take effect automatically.
In the future if we define `DeveloperNetworkPolicy`, this will come even more handy
for uniformity, especially across `Pass` actions.
```